source/iopipe/json/serialize.d

```diff
@@ -845,118 +836,511 @@ unittest
     assert(cArr.length == 3);
 }
 
-void serializeImpl(T, Char)(scope void delegate(const(Char)[]) w, ref T val) if (__traits(isStaticArray, T))
+//alias OutputDelegate(Char) = void delegate(const(Char)[]);
+
+struct FormattingOptions
+{
+    bool lineBreaks;
+    bool spaceAfterColon;
+    private bool _spaceBetweenItems;
+    // Only makes sense if lineBreaks == false
+    bool spaceBetweenItems() => !lineBreaks && _spaceBetweenItems;
+    bool spaceBetweenItems(bool value)
+    {
+        _spaceBetweenItems = value;
+        return spaceBetweenItems;
+    }
+    bool trailingCommas;
+    bool preferUnquotedKeys;
+    bool preferSingleQuotes;
+    string quoteText() => preferSingleQuotes ? "'" : "\"";
+    private bool _indentation;
+    // Only makes sense if lineBreaks == true
+    bool indentation() => lineBreaks && _indentation;
+    bool indentation(bool value)
+    {
+        _indentation = value;
+        return indentation;
+    }
+    int indentationLevel;
+    string indentationText;
+    
+    FormattingOptions indent()
+    {
+        FormattingOptions fo = this;
+        fo.indentationLevel++;
+        return fo;
+    }
+}
+
+/// Returns:
+///     Whether the string is a valid ECMA 5.1 identifier name.
+///     Used for checking if JSON5 key can be written without quotes.
+bool isEcmaIdentifierName(S)(S str)
```

Comment:
I think you can just use std.uni.isAlpha instead of this. And it already does all the things you do below I believe.

Reply:
No, actually, that's just the token, which doesn't validate the name. The validation is done later, but that's just an `isAlphaNum`.

Technically, JSON5 doesn't allow JS keywords, but I ignore that requirement. So isAlphaNum works fine.

Reply:
[JSON5 spec](https://spec.json5.org) says "Object keys may be an ECMAScript 5.1 [IdentifierName](https://www.ecma-international.org/ecma-262/5.1/#sec-7.6)." and ECMAScript 5.1 spec states "An *Identifier* is an *IdentifierName* that is not a *ReservedWord*" so JSON5 does in fact allow keywords because it uses *IdentifierName*.

As for `std.uni.isAlphaNum`, I don't think it does the same thing here. isEcmaIdentifierName returns true for `"idẽn̈tifier‿٩٨٣٤"` but isAlphaNum does not. ECMAScript 5.1 spec allows categories like connector punctuation (Pc), non-spacing mark (Mn) etc.

Reply:
Those `isSomething()` functions in phobos are not specific enough. See: [Unicode Character Categories](https://www.compart.com/en/unicode/category). There are 7 different sub-classes for "Punctuation" and yet phobos only provides `isPunctuation()` which does not differentiate between them, no function named `isConnectorPunctuation()` exists.

We're not reimplementing much here though, we just get the `CodepointSet` from `std.uni` and check if the character from the json string belongs to it.

Reply:
OK, I'm really confused at the code for std.uni. There seem to be two ways that characters are checked -- one is via a trie, and then there is this mechanism, which appears to decompress and initialize the tables at runtime.

But yeah, you are right that we need a mechanism for this. I wonder if it would be worth doing the trie path with all the things we need here? That should do everything at compile time, and just be one lookup instead of multiple.

Reply:
Yeah, I think that's what I want to do. I want to do whatever isAlphaNum does, but for JSON5 symbols. Then it's just a compile-time lookup table. Added #10 to track this.

One thing that we need to make sure is that this can be used at compile time (though I don't think I've tested it). Added #11 to track this